### ISSUE TITLE
[refactor] use termcolor crate instead of ansi

### ISSUE THREAD

**Issue by HerringtonDarkholme** [ast-grep/ast-grep#156]
Using ansi will make some windows client unhappy because cygwin like terminals does not support ansi but using console api.

**Comment by pd4d10** [comment]
Is this refactor started? If not I'd like to take it

**Comment by HerringtonDarkholme** [comment]
Nope, it is not started. Happy to see you are interested but would you like to provide a high-level plan for the refactor?
It will be quite huge.

**Comment by pd4d10** [comment]
Sure, I'll look into it.

**Comment by pd4d10** [comment]
After doing some research, I found that the usage of `termcolor` is a little different from other libraries (with just ANSI support), because [it aims to support the Windows console API](https://internals.rust-lang.org/t/auditing-terminal-text-color-crates-usages/10158/6), which also means it seems difficult to log texts with different styles in one same statement. For example:

```rs
// the styles of header and message are different
writeln!(w, "{header} {message}");
```

should be firstly refactored into

```rs
write!(w, header);
writeln!(w, " {message}");
```

So I think the refactor could be split into two steps:
1. Following the example above, refactor all the existing print stuff.
2. Finish the `ansi_term -> termcolor` replacement.

The trade-off is that we would lose some intuitiveness when doing print stuff in the future.

What do you think?

**Comment by HerringtonDarkholme** [comment]
Nice writeup! I wonder if we can test the huge refactor accordingly? If the end user is not impacted by the change, I would opt for termcolor since it also supports windows console API.

**Comment by pd4d10** [comment]
> if we can test the huge refactor accordingly

Yeah, we can firstly improve the test coverage, which is currently 60% according to the badge at README, then do the refactor.

**Comment by pd4d10** [comment]
It seems that we already have tests, but there is no check for the color. For example

https://github.com/ast-grep/ast-grep/blob/d9ebbe5cc2b47876c8ef212550f34891b64bfd28/crates/cli/src/error.rs#L254-L289

I guess we can use snapshot to avoid hard code such as `"\u1b[31"` in the test cases. A lib named [insta](https://github.com/mitsuhiko/insta) seems to meet our needs.

Shall we use it or just do the hard code in test cases, what do you think?

**Comment by RReverser** [comment]
Would this also help with fixing color support on Windows? Currently `--color` says it default to `auto` which is described as

>           - auto:   Try to use colors, but don't force the issue. If the output is piped to another program, or the console isn't available on Windows, or if TERM=dumb, or if `NO_COLOR` is defined, for example, then don't use colors

but I don't know what "console isn't available on Windows" means. I'm on Windows 11 and I always have to manually provide `sg --color ansi` to get colors in Windows Terminal as well as VSCode terminal, which gets a bit annoying to have to remember.

**Comment by HerringtonDarkholme** [comment]
> Would this also help with fixing color support on Windows? 

I do know if using termcolor should somehow fix Windows usage. Windows Terminal and VSCode terminal accepts ansi color so `--color ansi`. 

I believe termcolor supports Windows console API which supports color natively without using ANSI escape code. Unfortunately, I don't have Windows machine to test the behavior of termcolor :( 

I would try to refactor usage in future so external contributors can test termcolor.

**Comment by RReverser** [comment]
> Windows Terminal and VSCode terminal accepts ansi color

More specifically, it's accepted by Windows itself since Windows 10, not by individual apps. Unless you want to support old Windows versions (which hopefully shouldn't be necessary for developers), it makes just defaulting to ANSI colors a much easier cross-platform option.

**Comment by RReverser** [comment]
There is some Windows API you'll still need to check if colors are supported (e.g. to make sure it's not a redirect or something), but I believe one of these crates should already handle that for you.

**Comment by HerringtonDarkholme** [comment]
Sounds good. I will investigate it. Changing default color is tracked by #680

**Comment by HerringtonDarkholme** [comment]
It is supported in 0.13.1!

**Comment by RReverser** [comment]
Thanks!
